clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -242,46 +317,36 @@ public PrototypeAsyncConsumer(LogContext logContext,
 
-            do {
-                if (!eventHandler.isEmpty()) {
-                    final Optional<BackgroundEvent> backgroundEvent = eventHandler.poll();
-                    // processEvent() may process 3 types of event:
-                    // 1. Errors
-                    // 2. Callback Invocation
-                    // 3. Fetch responses
-                    // Errors will be handled or rethrown.
-                    // Callback invocation will trigger callback function execution, which is blocking until completion.
-                    // Successful fetch responses will be added to the completedFetches in the fetcher, which will then
-                    // be processed in the collectFetches().
-                    backgroundEvent.ifPresent(event -> log.warn("Do something with this background event: {}", event));
-                }
+            if (this.subscriptions.hasNoSubscriptionOrUserAssignment()) {
+                throw new IllegalStateException("Consumer is not subscribed to any topics or assigned any partitions");
+            }
 
-                updateAssignmentMetadataIfNeeded();
+            do {
+                updateAssignmentMetadataIfNeeded(timer);
+                final Fetch<K, V> fetch = pollForFetches(timer);
 
-                // Create our event as a means to request the background thread to return any completed fetches.
-                // If there are any
-                Queue<CompletedFetch<K, V>> completedFetches = eventHandler.addAndGet(new FetchEvent<>(), timeout);
+                if (!fetch.isEmpty()) {
+                    sendFetches();
 
-                if (completedFetches != null && !completedFetches.isEmpty()) {
-                    fetchBuffer.addAll(completedFetches);
-                }
+                    if (fetch.records().isEmpty()) {
+                        log.trace("Returning empty records from `poll()` "
+                                + "since the consumer's position has advanced for at least one topic partition");
+                    }
 
-                // The idea here is to have the background thread sending fetches autonomously, and the fetcher
-                // uses the poll loop to retrieve successful fetchResponse and process them on the polling thread.
-                final Fetch<K, V> fetch = fetchCollector.collectFetch(fetchBuffer);
-                if (!fetch.isEmpty()) {
                     return this.interceptors.onConsume(new ConsumerRecords<>(fetch.records()));
                 }
                 // We will wait for retryBackoffMs
             } while (timer.notExpired());
-        } catch (final Exception e) {
-            throw new RuntimeException(e);
-        }
 
-        return ConsumerRecords.empty();
+            return ConsumerRecords.empty();
+        } finally {
+            this.kafkaConsumerMetrics.recordPollEnd(timer.currentTimeMs());
+        }
     }
 
     boolean updateAssignmentMetadataIfNeeded() {
```

Comment:
it seems like this is not referenced anywhere @lianetm do we want to delete this?

Reply:
this branch it outdated, `updateAssignmentMetadataIfNeeded` is already integrated into the poll loop so should be fine when we merge this.